### PR TITLE
Initial Support for v2.0.0

### PR DESCRIPTION
**Pull request by zshaheen**
* Refactored lots of things, so it's much easier for anyone to add new diagnostic **sets**.
  * We're merging into master so others can begin developing for `e3sm_diags`.
  * Documentation regarding this will be done by Monday.
* Initial support for the Area Mean Time Series plots.
  * They just need some polishing and bug-fixes.
  * Before the release, these should be done.
* Added the ability to run `e3sm_diags` programmatically via an API.

### CODE REVIEW

**Comment by zshaheen** [comment]
Also, we've changed how diagnostics are run. You use the API to run, as per the example below. We will have documentation explaining this, as well as converting all examples.

**However, there is backwards compatibility, so you can still run `e3sm_diags` as you normally would.** Only the current plot sets will be supported running the "old way", since the newer plots, like `area_mean_time_series`, have set-specific parameters and you need to run it the "new way".

---

Here's [a parameters script, `myparams.py`, from the documentation](https://e3sm-project.github.io/e3sm_diags/docs/html/quickguides/quick-guide-cori.html#running-the-entire-annual-latitude-longitude-contour-set)
```python
reference_data_path = '/global/project/projectdirs/acme/acme_diags/obs_for_e3sm_diags/climatology/'
test_data_path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/climatology/'

test_name = '20161118.beta0.FC5COSP.ne30_ne30.edison'

sets = ["lat_lon"]
seasons = ["ANN"]

# 'mpl' and 'vcs' are for matplotlib or vcs plots respectively.
backend = 'mpl'

# Name of folder where all results will be stored.
results_dir = 'lat_lon_demo'
```

You run it like so:
```
e3sm_diags -p myparams.py
```

---

Now with to run with the API, you create a script, `run_something.py`:
```
from acme_diags.parameter.core_parameter import CoreParameter
from acme_diags.run import runner

param = CoreParameter()

param.reference_data_path = '/global/project/projectdirs/acme/acme_diags/obs_for_e3sm_diags/climatology/'
param.test_data_path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/climatology/'
param.test_name = '20161118.beta0.FC5COSP.ne30_ne30.edison'
param.seasons = ["ANN"]

prefix = '/global/project/projectdirs/acme/www/shaheen2/runs_with_api'
param.results_dir = os.path.join(prefix, 'lat_lon_demo')

runner.sets_to_run = ['lat_lon']
runner.run_diags([param])
```

And you run it like so:
```
python run_something.py
```

**Comment by chengzhuzhang** [comment]
@zshaheen Great progress!
I tested the branch, it ran well. I did find two problems:
1. `area_mean_time_series` is not shown up in the viewer, when I tested all sets. Here is my [test run](https://portal.nersc.gov/project/acme/zhang40/runs_with_api/test/viewer/)

2. A small thing:
In acme_diags_driver, the used driver files are printed multiple times, as following:
```
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/zonal_mean_xy/zonal_mean_xy_model_vs_obs.cfg for zonal_mean_xy.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/zonal_mean_2d/zonal_mean_2d_model_vs_obs.cfg for zonal_mean_2d.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/meridional_mean_2d/meridional_mean_2d_model_vs_obs.cfg for meridional_mean_2d.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/lat_lon/lat_lon_model_vs_obs.cfg for lat_lon.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/polar/polar_model_vs_obs.cfg for polar.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/area_mean_time_series/area_mean_time_series_model_vs_obs.cfg for area_mean_time_series.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/cosp_histogram/cosp_histogram_model_vs_obs.cfg for cosp_histogram.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/zonal_mean_xy/zonal_mean_xy_model_vs_obs.cfg for zonal_mean_xy.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/zonal_mean_2d/zonal_mean_2d_model_vs_obs.cfg for zonal_mean_2d.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/meridional_mean_2d/meridional_mean_2d_model_vs_obs.cfg for meridional_mean_2d.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/lat_lon/lat_lon_model_vs_obs.cfg for lat_lon.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/polar/polar_model_vs_obs.cfg for polar.
Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/area_mean_time_series/area_mean_time_series_model_vs_obs.cfg for area_mean_time_series.
```
I suspect get_parameter function was called more than it should be.

File: acme_diags/acme_diags_driver.py [hunk 1]
```diff
@@ -124,13 +124,50 @@ def _save_parameter_files(results_dir, parser):
...
+
+
 def save_provenance(results_dir, parser):
     """
     Store the provenance in results_dir.
     """
     results_dir = os.path.join(results_dir, 'prov')
     if not os.path.exists(results_dir):
-        os.makedirs(results_dir, 0o775)
+        os.makedirs(results_dir, 0o755)
```
Comment (chengzhuzhang) [inline acme_diags/acme_diags_driver.py]: Very thoughtful to set permission for result_dir!

File: acme_diags/metrics/__init__.py [hunk 2]
```diff
@@ -1,14 +1,47 @@
...
-from . import min_cdms
-from . import std
-
-corr = corr.CORR()
-rmse = rmse.RMSE()
-max_cdms = max_cdms.Max()
-mean = mean.Mean()
-min_cdms = min_cdms.Min()
-std = std.STD()
+import numpy
```
Comment (chengzhuzhang) [inline acme_diags/metrics/__init__.py]: I like it much better to have metrics calculation as functions in one file as modified.

**Comment by zshaheen** [comment]
> @zshaheen Great progress!
> I tested the branch, it ran well. I did find two problems:
> 
> 1. `area_mean_time_series` is not shown up in the viewer, when I tested all sets. Here is my [test run](https://portal.nersc.gov/project/acme/zhang40/runs_with_api/test/viewer/)
> 2. A small thing:
>    In acme_diags_driver, the used driver files are printed multiple times, as following:
> 
> ```
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/zonal_mean_xy/zonal_mean_xy_model_vs_obs.cfg for zonal_mean_xy.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/zonal_mean_2d/zonal_mean_2d_model_vs_obs.cfg for zonal_mean_2d.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/meridional_mean_2d/meridional_mean_2d_model_vs_obs.cfg for meridional_mean_2d.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/lat_lon/lat_lon_model_vs_obs.cfg for lat_lon.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/polar/polar_model_vs_obs.cfg for polar.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/area_mean_time_series/area_mean_time_series_model_vs_obs.cfg for area_mean_time_series.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/cosp_histogram/cosp_histogram_model_vs_obs.cfg for cosp_histogram.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/zonal_mean_xy/zonal_mean_xy_model_vs_obs.cfg for zonal_mean_xy.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/zonal_mean_2d/zonal_mean_2d_model_vs_obs.cfg for zonal_mean_2d.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/meridional_mean_2d/meridional_mean_2d_model_vs_obs.cfg for meridional_mean_2d.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/lat_lon/lat_lon_model_vs_obs.cfg for lat_lon.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/polar/polar_model_vs_obs.cfg for polar.
> Using /global/homes/c/chengzhu/.conda/envs/e3sm_diags_env_dev_jun12/share/e3sm_diags/area_mean_time_series/area_mean_time_series_model_vs_obs.cfg for area_mean_time_series.
> ```
> 
> I suspect get_parameter function was called more than it should be.

Regarding point 1, that's because `area_mean_time_series` only works with time-series data as input. We're using climo files in that example that you ran. An error is printed however, but it's probably hard to notice.

In terms of point 2, I've noticed that as well. To make things easier, in terms of the code and design (and not because I want to take the lazy approach), I'll just **not** have the function print what `cfg` files are used. Given the `run_type` and set, it's pretty easy to figure out what file is used. And most people probably aren't even going to look at those files anyways.

File: acme_diags/metrics/__init__.py (hunk 2 above)
Comment (zshaheen) [reply acme_diags/metrics/__init__.py]: @chengzhuzhang, so do you like it the new way, or the older way where we had separate files?

Comment (chengzhuzhang) [reply acme_diags/metrics/__init__.py]: Like the new way:)

**Comment by chengzhuzhang** [comment]
Okay, Thanks! Let me test the time-series setting for area mean time series...

**Comment by chengzhuzhang** [comment]
The time-series results (here)[https://portal.nersc.gov/project/acme/zhang40/runs_with_api/test_ts/viewer/]. Only one thing, relating to time slice, is problematic. I used year 2003 to 2006, how ever the test data only available to 2005. The time labeled on lat-lon map is still (2003 to 2006) for the area mean time series plot, the model only have two years data point while reference data have 4 data points. it's probably not a easy fix. I'm just leaving a note here. we can fix this at a later time.

**Pull request merged**